### PR TITLE
Storing comments in a commentmap

### PR DESCRIPTION
**Pull request by wolfgarnet**
This change will add a comment map to the parser, from where parsed comments are stored, indexed by nodes.
It will solve issue #148

### CODE REVIEW

**Comment by deoxxa** [comment]
At a glance, I'm quite impressed with the test coverage! It'll probably take a little while to read through the implementation, so please bear with us - I or someone else will probably have some questions over the next few days :)

**Comment by wolfgarnet** [comment]
Thank you. I will be ready to answer then.

File: ast/comments.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+			// Remove the comment from the "from" slice
+			cm[from][i] = cm[from][len(cm[from])-1]
+			cm[from][len(cm[from])-1] = nil
+			cm[from] = cm[from][:len(cm[from])-1]
+		}
+	}
+}
+
+// Display displays the map
+func (cm CommentMap) Display() {
```
Comment (darkliquid) [inline ast/comments.go]: This strikes me as something that doesn't necessarily belong on otto core. Developers should use a CommentMap and process it/display it in their own way.

**Comment by darkliquid** [comment]
I've not built this and played with it yet, but reading through it, it looks good. ANy chance you might be able to polish off that trailing comma comment TODO (with a test) soon? If so, if this gets the all clear it would be good to get that in before merging this.

File: ast/comments.go (hunk 1 above)
Comment (wolfgarnet) [inline ast/comments.go]: I agree. I was using it for debugging.

**Comment by wolfgarnet** [comment]
I will try to do it over the holidays. 
For the record, code will be parsed as usual, but comments will just not be stored.

**Comment by wolfgarnet** [comment]
I have added empty expressions for empty array elements instead of nils. Including tests.

**Comment by joushou** [comment]
Any progress on the review? I would find this PR useful as well.

**Comment by darkliquid** [comment]
I'll try and find the time to have another read through and play with the PR this weekend, but someone else will need to merge it if it's approved.

**Comment by darkliquid** [comment]
I'm pretty happy with this, I can't really point out anything I'd have a problem with and it all looks good to me.

:+1:

File: ast/comments.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return "Leading"
+	case TRAILING:
+		return "Trailing"
+	case KEY:
+		return "Key"
+	case COLON:
+		return "Colon"
+	case FINAL:
+		return "Final"
+
```
Comment (stevenh) [inline ast/comments.go]: unneeded blank line

File: ast/comments_test.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+func TestCommentMap_move(t *testing.T) {
+	statement1 := &EmptyStatement{file.Idx(1)}
+	statement2 := &EmptyStatement{file.Idx(2)}
+	comment := &Comment{1, "test", LEADING}
+
+	cm := CommentMap{}
+	cm.AddComment(statement1, comment)
+
+	if cm.Size() != 1 {
+		t.Errorf("The number of comments is %v, not 1", cm.Size())
```
Comment (stevenh) [inline ast/comments_test.go]: Errors should be lower case, more below

File: parser/comments_test.go [hunk 4]
```diff
@@ -0,0 +1,1296 @@
...
+import (
+	"fmt"
+	"github.com/robertkrimen/otto/ast"
+	"reflect"
+	"testing"
+)
+
+func checkComments(actual []*ast.Comment, expected []string, position ast.CommentPosition) error {
+	if len(actual) != len(expected) {
+		return fmt.Errorf("The number of comments is not correct. %v != %v", len(actual), len(expected))
```
Comment (stevenh) [inline parser/comments_test.go]: Errors should be lower case, move in other locations

File: parser/comments_test.go [hunk 5]
```diff
@@ -0,0 +1,1296 @@
...
+		parser, program = test(`
+for(var i = 0 in  obj) {
+	a
+}
+// comment
+	`, nil)
+		is(parser.commentMap.Size(), 1)
+		is(checkComments((*parser.commentMap)[program], []string{" comment"}, ast.FINAL), nil)
+
+		//		// Block
```
Comment (stevenh) [inline parser/comments_test.go]: Looks like some old / debug left in?

File: parser/parser.go [hunk 6]
```diff
@@ -270,3 +280,61 @@ func (self *_parser) position(idx file.Idx) file.Position {
...
+// findComments finds the following comments.
+// Comments on the same line will be grouped together and returned.
+// After the first line break, comments will be added as statement comments.
+func (self *_parser) findComments(ignoreLineBreak bool) []*ast.Comment {
+	comments := make([]*ast.Comment, 0)
+
+	newline := false
+
+	for self.implicitSemicolon == false || ignoreLineBreak {
+
```
Comment (stevenh) [inline parser/parser.go]: Unneeded blank line

File: parser/statement.go [hunk 7]
```diff
@@ -218,12 +262,21 @@ func (self *_parser) parseFunctionBlock(node *ast.FunctionLiteral) {
...
 	node := &ast.DebuggerStatement{
 		Debugger: idx,
 	}
 
+	self.commentMap.AddComments(node, comments, ast.TRAILING)
+
 	self.semicolon()
 
+	if !self.skippedLineBreak {
+		comments = self.findComments(false)
```
Comment (stevenh) [inline parser/statement.go]: any reason for assign instead of direct use?

**Comment by stevenh** [comment]
Given the amount of code and calls involved, what's the performance impact of this on a large script?

File: parser/statement.go (hunk 7 above)
Comment (wolfgarnet) [inline parser/statement.go]: I just try not to have too many function calls in one line. I'm not religious about it, though.

File: parser/comments_test.go (hunk 4 above)
Comment (wolfgarnet) [inline parser/comments_test.go]: I have lower cased my errors, but I have not moved them. Yet. The errors used in tests, should they be moved? If I only use an error once, I don't think it should be moved.

**Comment by wolfgarnet** [comment]
No I have not performance tested it. I could give it a shot later today or tomorrow.

**Comment by wolfgarnet** [comment]
Has otto been performance tested before? Are there any code I can reuse?

**Comment by stevenh** [comment]
I'm not aware of it, but this could easily be done using go's benchmarking tests.

If this does add noticeable delays it would be good to add a flag to enable comment parsing, which defaults to disabled.

**Comment by wolfgarnet** [comment]
I will return when I have some numbers.

**Comment by wolfgarnet** [comment]
I have some numbers from the added benchmark. It parses the underscore source and finds 243 comments in all.
Master:           3091912ns/op ~ 3.09ms/op
commentmap: 3866777ns/op ~ 3.87ms/op
Which roughly is a slow down of 25%

Is that too much?

**Comment by wolfgarnet** [comment]
Wrong button, sorry!

**Comment by stevenh** [comment]
Yer I think that's significant enough for it to be disabled by default tbh as its only really of use in as pre-processor which is not the typical use case.

**Comment by wolfgarnet** [comment]
Ok. I will figure out a way to disable the comment parsing by default.

**Comment by wolfgarnet** [comment]
I have added a mode to toggle comment storing. It halves the performance penalty to about 13% overhead.

**Comment by wolfgarnet** [comment]
I have improved performance significantly. It's a bit more code, though. The penalty is now 2.6%.

File: parser/expression.go [hunk 8]
```diff
@@ -196,11 +200,18 @@ func (self *_parser) parseVariableDeclarationList(var_ file.Idx) []ast.Expressio
 	var list []ast.Expression
 
 	for {
-		list = append(list, self.parseVariableDeclaration(&declarationList))
+		comments := self.findComments(false)
+
+		decl := self.parseVariableDeclaration(&declarationList)
+		if self.mode & StoreComments != 0 { self.commentMap.AddComments(decl, comments, ast.LEADING) }
```
Comment (stevenh) [inline parser/expression.go]: Don't test twice and avoid trailing blocks.

Have you run gofmt on this as I didn't think it allowed trailing blocks?

Comment (wolfgarnet) [inline parser/expression.go]: Sorry. I will fix it.

**Comment by wolfgarnet** [comment]
Now the code is formatted.

**Comment by wolfgarnet** [comment]
Is there anything else I can do to get it merged?

**Comment by stevenh** [comment]
Looks like only some of the files were run through fmt e.g. parser/lexer.go seems to still have issues.

**Comment by wolfgarnet** [comment]
Sorry, I must have missed it.

**Comment by wolfgarnet** [comment]
I'm so sorry for the inconvenience, but could you have another look at it now?

**Comment by joushou** [comment]
Was gofmt the only issue remaining for this PR?

**Comment by darkliquid** [comment]
It was as far as I was concerned. I'm not sure if @stevenh had any other concerns.

**Comment by wolfgarnet** [comment]
Could you please explain to me, what I have to do to get it merged?

**Comment by deoxxa** [comment]
This looks good - thanks for sticking with it, @wolfgarnet!

**Comment by wolfgarnet** [comment]
Thank you deoxxa.

**Pull request merged**